Add `esp8266_disable_ssl_support:` config option

ayufan opened the pull request:
# What does this implement/fix? 

ESP8266 still does come in 1MB flavor which very often is not enough to have a support
for HTTPS (for `http_request`) and `OTA`.

This allows to disable `SSL` support to have smaller compilation size as very often
no SSL support is required.

This is especially useful when used with `http_request` which greatly
inflates binary size due to implicit SSL support. The ESP32 has significantly
more space and does not really need this option.

## Outcome

### With `esp8266_disable_ssl_support:` not configured or set to `false`

```bash
Checking size .pioenvs/mk3s_switch/firmware.elf
RAM:   [====      ]  43.4% (used 35532 bytes from 81920 bytes)
Flash: [======    ]  58.7% (used 600888 bytes from 1023984 bytes)
...
INFO Uploading configs-iot/mk3s_switch/.pioenvs/mk3s_switch/firmware.bin (605040 bytes)
ERROR Error binary size: Error: ESP does not have enough space to store OTA file. Please try flashing a minimal firmware (remove everything except ota)
```

### With `esp8266_disable_ssl_support: true`

```bash
Checking size .pioenvs/mk3s_switch/firmware.elf
RAM:   [====      ]  42.9% (used 35124 bytes from 81920 bytes)
Flash: [=====     ]  48.1% (used 492744 bytes from 1023984 bytes)
...
INFO Uploading configs-iot/mk3s_switch/.pioenvs/mk3s_switch/firmware.bin (496896 bytes)
...
```

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/1443

## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
esphome:
  board: esp01_1m

http_request:
  useragent: esphome/device
  timeout: 3s
  esp8266_disable_ssl_support: true
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

ayufan committed:
Add esp8266_disable_ssl_support config option

ESP8266 still does come in 1MB flavor which
very often is not enough to have a support
for HTTPS (for `http_request`) and `OTA`.

This allows to disable `SSL` support to have
smaller compilation size as very often
no SSL support is required.

oxan reviewed (commented):
So, I like the idea, but a couple notes on the implementation

* Our feature flags usually start with `USE_` instead of `ENABLE_`, and I think it'd be nice to indicate that this flag is specifically about HTTP requests (so e.g. `USE_ESP8266_HTTPS_CLIENT`).
* We define feature flags using `cg.add_define()` instead of a compiler flag (so that files that don't depend on the feature flags don't have to be recompiled when they are changed).
* The flag should also be added to `esphome/core/defines.h` so that the disabled code is still CI-tested.
* Can we move the setting to be under the `http_request` component instead of the `esphome` component?

ayufan commented:
Thanks @oxan a few responses about choices:

> Our feature flags usually start with USE_ instead of ENABLE_, and I think it'd be nice to indicate that this flag is specifically about HTTP requests (so e.g. USE_ESP8266_HTTPS_CLIENT).

I was thinking about `USE_` but I did not find any good name for that. Any suggestions?

> We define feature flags using cg.add_define() instead of a compiler flag (so that files that don't depend on the feature flags don't have to be recompiled when they are changed).

Maybe I missed and likely it needs to be added to `http_request` and `nexion`, and it was not working properly for me without a `build_arg`. However, if this is added to `defines.h` and then to `http_request.h` and `nextion/nextion.h` it should work likely.

> Can we move the setting to be under the http_request component instead of the esphome component?

I was thinking about that as well, but decided against of that:
- there are two users: `http_request` and `nextion`: this covers two
- this seems to work well with the prior `esp8266_*` specific option that was in `core:` context
- this is needed only only for some very limited boards, so inflating `http_request` with additional options might not be desirable, but I don't feel strongly about that

ayufan commented:
@oxan 

I changed everything except `http_request:` since this feels more like a global setting, and since this also affects `nextion` integration as well.

OttoWinter commented:
> I changed everything except http_request: since this feels more like a global setting

That should be changed back; the setting should be specific to the components using SSL (even if there are multiple)

ayufan commented:
@OttoWinter Any sugestion about config name?

OttoWinter commented:
> Any suggestion about config name, or is the name fine?

`esp8266_disable_ssl_support` is fine with me

The `USE_` define should be prefixed with the name of the integration though (as those are global), like `USE_HTTP_REQUEST_ESP8266_HTTPS`

ayufan commented:
@OttoWinter 

As you noticed this also supports today the `esphome/components/nextion/nextion_upload.cpp` opening an SSL connection. After the `http_request:` focused change I will remove this `#ifdef` in this context. I assume this is what you expect and you consider that we might add a separate option there in the future and in all places using ssl sockets.

Asking to be sure :)

ayufan commented:
@OttoWinter Updated and moved into `http_request:`. I also validated on my SP111 plug (ESP8266). Docs are also updated https://github.com/esphome/esphome-docs/pull/1443. The docs looks like this: https://deploy-preview-1443--esphome.netlify.app/components/http_request.html#esphome-esp8266-disable-ssl-support.

oxan commented on esphome/components/http_request/__init__.py:
```diff
@@ -98,6 +102,8 @@ async def to_code(config):
     var = cg.new_Pvariable(config[CONF_ID])
     cg.add(var.set_timeout(config[CONF_TIMEOUT]))
     cg.add(var.set_useragent(config[CONF_USERAGENT]))
+    if not config.get(CONF_ESP8266_DISABLE_SSL_SUPPORT, False):
```
```suggestion
    if not config[CONF_ESP8266_DISABLE_SSL_SUPPORT]:
```
No need to repeat the default here.

ayufan replied:
~Fixed.~

ayufan replied:
Unfortunately it does not work for `esp32`, since in this context it is not considered to be defined by defualt, thus why we do `.get(..., False)`. I will revert it back. I see tests for `tests/test1.yaml` to break.

oxan replied:
Oh, you're right about that. I would instead recommend this then, to avoid creating the define on ESP32 as well (not that it hurts, but it's nice to be clean):
```suggestion
    if CORE.is_esp8266 and not config[CONF_ESP8266_DISABLE_SSL_SUPPORT]:
```

ayufan replied:
I'm slightly tired of changing this over and over :)

oxan commented on esphome/components/nextion/nextion.h:
```diff
@@ -774,7 +776,9 @@ class Nextion : public NextionBase, public PollingComponent, public uart::UARTDe
 #if defined(USE_ETHERNET) || defined(USE_WIFI)
 #ifdef ARDUINO_ARCH_ESP8266
   WiFiClient *wifi_client_{nullptr};
+#ifdef USE_HTTP_REQUEST_ESP8266_HTTPS
```
This flag is only defined if the `http_request` component is loaded, and nextion doesn't autoload it; thus HTTPS support here will now be disabled by default.

ayufan replied:
Thanks. Removed `nextion` changes.

ayufan committed:
Remove nextion changes

ayufan committed:
Use CORE.is_esp8266 and not... syntax

ayufan commented:
@OttoWinter Would you have a moment to take a look again?

OttoWinter reviewed (approved):
Thanks!
